server/http/src/service.rs

```diff
@@ -37,19 +70,31 @@ impl GraphQLService {
     }
 
     /// Handles GraphQL queries received via POST /.
-    fn handle_graphql_query(&self, request: Request<Body>) -> GraphQLServiceResponse {
+    fn handle_graphql_query(
+        &self,
+        name_or_id: &str,
+        request: Request<Body>,
+    ) -> GraphQLServiceResponse {
         let query_sink = self.query_sink.clone();
-        let schema = self.schema.clone();
+        let schemas = self.schemas.read().unwrap();
+
+        // First try `name_or_id` as a name, if that fails try it as an id.
+        let schema = schemas
+            .get(name_or_id)
+            .or_else(|| schemas.get(self.names.read().unwrap().get(name_or_id)?));
+
+        let schema = if let Some(schema) = schema {
+            schema.clone()
+        } else {
+            return self.handle_not_found();
+        };
 
         Box::new(
             request
                 .into_body()
                 .concat2()
                 .map_err(|_| GraphQLServerError::from("Failed to read request body"))
-                .and_then(move |body| {
-                    let schema = schema.lock().unwrap();
-                    GraphQLRequest::new(body, schema.clone())
-                })
+                .and_then(move |body| GraphQLRequest::new(body, schema.clone()))
```

Comment:
I wonder: since we're already cloning the schema in line 86, is this clone necessary?

Reply:
Yes, because the closure owns `schema` and if it moves it into the `GraphQLRequest` when called then it'd be a `FnOnce` not an `Fn`.

Reply:
Got it!